x/swaprouter/creator.go

```diff
@@ -0,0 +1,124 @@
+package swaprouter
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/pkg/errors"
+
+	"github.com/osmosis-labs/osmosis/v12/osmoutils"
+	gammtypes "github.com/osmosis-labs/osmosis/v12/x/gamm/types"
+	"github.com/osmosis-labs/osmosis/v12/x/swaprouter/types"
+)
+
+// CreatePool attempts to create a pool returning the newly created pool ID or
+// an error upon failure. The pool creation fee is used to fund the community
+// pool. It will create a dedicated module account for the pool and sends the
+// initial liquidity to the created module account.
+//
+// After the initial liquidity is sent to the pool's account, shares are minted
+// and sent to the pool creator. The shares are created using a denomination in
+// the form of gamm/pool/{poolID}. In addition, the x/bank metadata is updated
+// to reflect the newly created GAMM share denomination.
+func (k Keeper) CreatePool(ctx sdk.Context, msg types.CreatePoolMsg) (uint64, error) {
+	err := validateCreatePoolMsg(ctx, msg)
+	if err != nil {
+		return 0, err
+	}
+
+	sender := msg.PoolCreator()
+	initialPoolLiquidity := msg.InitialLiquidity()
+
+	// send pool creation fee to community pool
+	params := k.GetParams(ctx)
+	if err := k.communityPoolKeeper.FundCommunityPool(ctx, params.PoolCreationFee, sender); err != nil {
+		return 0, err
+	}
+
+	poolId := k.getNextPoolIdAndIncrement(ctx)
+	pool, err := msg.CreatePool(ctx, poolId)
+	if err != nil {
+		return 0, err
+	}
+
+	if err := k.validateCreatedPool(ctx, initialPoolLiquidity, poolId, pool); err != nil {
+		return 0, err
+	}
+
+	// create and save the pool's module account to the account keeper
+	if err := osmoutils.CreateModuleAccount(ctx, k.accountKeeper, pool.GetAddress()); err != nil {
+		return 0, fmt.Errorf("creating pool module account for id %d: %w", poolId, err)
+	}
+
+	swapModule := k.routes[msg.GetPoolType()]
+
+	if err := swapModule.InitializePool(ctx, pool, sender); err != nil {
+		return 0, err
+	}
+
+	// send initial liquidity to the pool
+	err = k.bankKeeper.SendCoins(ctx, sender, pool.GetAddress(), initialPoolLiquidity)
+	if err != nil {
+		return 0, err
+	}
+
+	k.poolCreationListeners.AfterPoolCreated(ctx, sender, pool.GetId())
+
+	return pool.GetId(), nil
+}
+
+// getNextPoolIdAndIncrement returns the next pool Id, and increments the corresponding state entry.
+func (k Keeper) getNextPoolIdAndIncrement(ctx sdk.Context) uint64 {
+	nextPoolId := k.GetNextPoolId(ctx)
+	k.setNextPoolId(ctx, nextPoolId+1)
+	return nextPoolId
+}
+
+func validateCreatePoolMsg(ctx sdk.Context, msg types.CreatePoolMsg) error {
+	err := msg.Validate(ctx)
+	if err != nil {
+		return err
+	}
+
+	initialPoolLiquidity := msg.InitialLiquidity()
+	numAssets := initialPoolLiquidity.Len()
+	if numAssets < gammtypes.MinPoolAssets {
+		return types.ErrTooFewPoolAssets
+	}
+	if numAssets > gammtypes.MaxPoolAssets {
```

Comment:
Let's move MinPoolAssets and MaxPoolAssets to the swaprouter types since they're not being used or called anywhere else in the gamm module